### ISSUE TITLE
[java] UnusedPrivateMethod FP with raw type missing from the classpath

### ISSUE THREAD

**Issue by wac84s** [pmd/pmd#5097]
**Affects PMD Version:** 
7.3.0

**Rule:** [UnusedPrivateMethod](https://docs.pmd-code.org/latest/pmd_rules_java_bestpractices.html#unusedprivatemethod)

**Description:**
private method referenced by method reference operators not getting detected as used

this::mymethod

**Code Sample demonstrating the issue:**

```java
package com.mytest;

import jakarta.validation.ConstraintViolation; //imported from jakarta.validation:jakarta.validation-api:3.0.2
import java.util.List;
import java.util.Set;

public class UnusedPrivateMethodFalsePositive {
    //this does not trigger UnusedPrivateMethod
    private void doWork(List obj) {
        obj.toString();
    }

    public void execute(Set<List<?>> listOfLists) {
        listOfLists.forEach(this::doWork);
    }

    //BUT this does???
    //UnusedPrivateMethod -  this as a false positive - but what is different?
    private void addValidationError(ConstraintViolation constraintViolation) {
        constraintViolation.toString();
    }

    public void addValidationErrors(Set<ConstraintViolation<?>> constraintViolations) {
        constraintViolations.forEach(this::addValidationError);
    }

}
```

**Expected outcome:**

It is not an unused private method

PMD reports a violation at line ..., but that's wrong. That's a false positive.

UnusedPrivateMethodFalsePositive.java:19:	UnusedPrivateMethod:	Avoid unused private methods such as 'addValidationError(ConstraintViolation)'.

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*

Maven AND cli both report it

**Comment by wac84s** [comment]
https://github.com/pmd/pmd/issues/5083 might be related.  Has use of method reference operator

**Comment by oowekyala** [comment]
I can't reproduce this. Your example does not compile. If I import `java.util.Set` or replace the Set by a List, everything's fine.

**Comment by wac84s** [comment]
Thanks updated it.   I was quickly shrinking my real class into a mock example.

**Comment by oowekyala** [comment]
I think you're right it shouldn't... Tbh since the 7.0.0 release these FPs have pointed out several bugs in the type resolution code and allowed us to improve it, which is why I have delayed fixing the rule...

**Comment by wac84s** [comment]
@oowekyala  i updated the above example.   I was trying to not add a dependency but it only fails with that for me anyway

**Comment by wac84s** [comment]
no familiar with this acronym "FPs". what is that?

**Comment by jsotuyod** [comment]
@wac84s FP is short for "false positive"

**Comment by adangel** [comment]
I can only reproduce it, when jakarta-validation-api is missing on the auxclasspath.

**Comment by oowekyala** [comment]
Surprisingly this has nothing to do with method references. The following also reproduces the bug:
```java
        public class UnusedPrivateMethodFalsePositive {
            private void doWork(List obj){}

            public void execute(List<?> listOfLists) {
                doWork(listOfLists);
            }

        }
```
Importantly `List` must be unresolved here. The bug is that for unresolved symbols `List<?>` is not considered a subtype of the raw `List`, when it should.

**Comment by sd-ebrukanat** [comment]
Following test also gives false positive

```
 
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.junit.jupiter.MockitoExtension;
import reactor.core.publisher.Mono;
import reactor.function.TupleUtils;
import java.util.Optional;
import static org.assertj.core.api.Assertions.assertThat;
@ExtendWith(MockitoExtension.class)
class PmdTest2 {
    @Test
    void test() {
        var res = Mono.zip(Mono.just(Optional.of("Some test")), Mono.just(TestClass.of("some 2")))
                .flatMap(TupleUtils.function((par1, par2) -> Mono.just(createProperty(par1, par2))))
                .map(objectProperty -> addRequired(objectProperty))
                .block();
        assertThat(res.required).isEqualTo("pop");
        assertThat(res.val).isEqualTo("Some test");
    }
    private TestClass createProperty(Optional<String> par1, TestClass objectProperty) {
        return TestClass.of(par1.orElse("par5"));
    }
    private TestClass addRequired(TestClass objectProperty) {
        return objectProperty.withRequiredProperty("pop");
    }
    private static class TestClass {
        private String required;
        private String val;
        TestClass(String val) {
            this.val = val;
        }
        public static TestClass of(String val) {
            return new TestClass(val);
        }
        public TestClass withRequiredProperty(String s) {
            required = s;
            return this;
        }
        public String getRequired() {
            return required;
        }
        public String getVal() {
            return val;
        }
    }
}
 
 ```

**Comment by oowekyala** [comment]
@sd-ebrukanat Please open a new ticket. This does not appear to be related to the original issue
